New component: Initial implementation of `KCard` component

AllanOXDi opened the pull request:
<!-- Please remove any unused sections -->

## Description
This PR introduces the KCard component  to KDS, It will act  a foundational building block for creating various card types within our product eco system. By serving as a base component for creating different card types like lesson cards, resource cards, and channel cards. 

#### Issue addressed

https://github.com/learningequality/kolibri-design-system/issues/530
https://github.com/learningequality/kolibri-design-system/issues/528
https://github.com/learningequality/kolibri-design-system/issues/529



## Changelog



-  [#625]
  - **Description:** Initial implementation of `KCard` component
  - **Products impact:** New Component
  - **Addresses:** [#530](https://github.com/learningequality/kolibri-design-system/issues/530)
  - **Components:** KCard
  - **Breaking:** No
  - **Impacts a11y:**  Yes



## Steps to test

I suggest looking at these issues 
https://github.com/learningequality/kolibri-design-system/issues/530
https://github.com/learningequality/kolibri-design-system/issues/528
https://github.com/learningequality/kolibri-design-system/issues/529

## (optional) Implementation notes

### At a high level, how did you implement this?
<!-- Briefly describe how this works -->

### Does this introduce any tech-debt items?
<!-- List anything that will need to be addressed later -->

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [x] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## After review

- [ ] The changelog item has been pasted to the `CHANGELOG.md`

## Comments
<!-- Any additional notes you'd like to add -->

AllanOXDi committed:
add kcard to kds

AllanOXDi committed:
add [data-focus=true] to handle card focus

AllanOXDi committed:
lint

AllanOXDi committed:
KCard prop validation and handling tabbing through

AllanOXDi committed:
added spacing to the image

AllanOXDi committed:
updates doc for KCard

AllanOXDi committed:
kcad documentation

AllanOXDi committed:
added prop linkage

AllanOXDi committed:
used semantic elements to wrap kcard content

AllanOXDi committed:
added both layouts for testing the kcard in the playground component

AllanOXDi committed:
updates doc

AllanOXDi committed:
clean the KCard doc

AllanOXDi commented:
I've added a few cards to the playground for @radinamatic  to perform QA testing. Once everything is approved by the QA team, we'll remove the cards before merging the code.

AllanOXDi committed:
lint

radinamatic commented:
Overall very good implementation, you followed all the recommended accessibility guidelines 👏🏽 💯 🙂 

One thing that needs to be corrected is the use of the semantic ``header`` and ``footer`` elements inside cards. We should have been more clear in the component specification, but the fact that we call something _card **footer**_ and _card **header**_ does not imply that semantic ``header`` and ``footer`` landmark elements need to be implemented. These are page-level elements and having them (announced) for each of potentially tens of cards on the page would be confusing.

Header | Footer
-- | --
<img width="683" alt="header" src="https://github.com/learningequality/kolibri-design-system/assets/1457929/70ab9601-7025-45e1-b164-6239652c1b7a"> | <img width="683" alt="footer" src="https://github.com/learningequality/kolibri-design-system/assets/1457929/c00a3982-ba93-423b-be73-b957a10ae5ae">

radinamatic commented:
Regarding the responsiveness, I don't know if it's because of Netlify view of the _playground_ (🙃) but I'm wondering if it's how it's supposed to look and behave...? Or maybe we can only properly test this once the cards are in the grid? 🤔 


https://github.com/learningequality/kolibri-design-system/assets/1457929/faa54f4a-6f99-4643-8967-0ad6248af5a0

radinamatic commented:
The last question I had was regarding the other interactive elements **inside** the card (button/icon to open other options, etc.): that is not in the scope of this PR and is supposed to come in subsequent ones, correct?

AllanOXDi commented:
> The last question I had was regarding the other interactive elements **inside** the card (button/icon to open other options, etc.): that is not in the scope of this PR and is supposed to come in subsequent ones, correct?


I think it's something we tried and tested here https://github.com/learningequality/studio/pull/4480

AllanOXDi commented:
> but I'm wondering if it's how it's supposed to look and behave...? Or maybe we can only properly test this once the cards are in the grid? 🤔

I think that's how it should behave for now.

> Regarding the responsiveness, I don't know if it's because of Netlify view of the _playground_ (🙃) but I'm wondering if it's how it's supposed to look and behave...? Or maybe we can only properly test this once the cards are in the grid? 🤔
> 
>  Untitled.Project.mp4

 The Grid will control it's behavior further . I will defer to @MisRob  for more thoughts and guidance on this.

radinamatic commented:
> I think it's something we tried and tested here [learningequality/studio#4480](https://github.com/learningequality/studio/pull/4480)

I don't recall having tested that, is it live somewhere for me to take a look?

AllanOXDi commented:
> I don't recall having tested that, is it live somewhere for me to take a look?

Not- it's not yet live for testing as dev is still on going.

MisRob commented:
Thank you @radinamatic and @AllanOXDi.

Regarding responsiveness - that's something that a grid will control, so yes, it is to be implemented and tested later on.

Regarding buttons in the footer area - `KCard` itself only provides place for them. It is products that will be responsible for adding buttons, context menus, etc. and ensuring it doesn't break a11y. We tried it a bit in a Studio PR as @AllanOXDi mentioned, but that's work in progress and there should be more thorough QA any time we add new cards in a product. After we start seeing some patterns, we may also consider adding some guidance to the documentation or adding features to `KCard` to further support such cases.

radinamatic commented:
All good then, will wait for the refactor of semantic `header` and` footer` landmark elements here to approve, and looking forward to testing the PRs for follow up steps of the new KCard! 🙂

AlexVelezLl commented on docs/tableOfContents.js:
```diff
@@ -402,6 +402,12 @@ export default [
         isCode: true,
       }),
+      new Page({
+        path: '/KCard',
```
The path needs to be lowercase `/kcard`. Otherwise the nuxt generate script wont catch the page:
![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/d280790a-b079-4f27-a879-818efac11e1c)

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+    <DocsPageSection title="Overview" anchor="#overview">
+      <p>
+        <code>KCard</code>
+        provides an accessible component for managing most frequently card content such as thumbnail,layouts and other customization options
```
We are missing a space here after `thumbnail,`, and a period at the end of the paragraph.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+      <h3>Vertical layout</h3>
+      <p>
+        Vertical
+        <DocsInternalLink href="/KCard#prop:layout">
```
In these props references we need to lowercase the path too.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+            :to="{ name: '/' }"
+            title="how to get started with Kolibri"
+            headingLevel="2"
+            layout="verticle"
```
Here we have a typo. This should be `layout="vertical"`

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+            layout="verticle"
+            thumbnailDisplay="none"
+          >
+            <template #title>
```
I didnt understand the role of this #title slot here. If we pass a title slot we dont need to pass the `title` prop? 🤔

AllanOXDi replied:
No- I was just trying with different slots - But I think it's confusing. Beside the title prop is not required.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+          >
+            <template #title>
+            </template>
+            <template #aboveTitle>
```
I dont see this aboveTitle above the title, I see it below :smiling_face_with_tear: :
![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/5b4be49a-5468-4212-b7ba-8c977e2d0ab9)

AllanOXDi replied:
I thinks it's something I just copied and pasted - will fix it.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        </docsinternallink> small
+      </h5>
+      <p>
+        This layout offers a combined view of content and a thumbnail image. This provides users with both textual information
```
I think we missed finishing this paragraph, maybe we wanted to put something like `This provides users with both textual information and visual representation.` or not?

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+          <KCard
+            :to="{ name: '/' }"
+            layout="vertical"
+            thumbnailDisplay="large"
```
Here we are showing an example of `thubmnailDisplay` small, but we are passing the value as "large".

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+       */
+      to: {
+        type: Object,
+        required: true,
```
IMO `to` prop should not be required, I think we could use cards just to show information or not? 🤔

AllanOXDi replied:
No- I think  prop `to` is required . Remember the entire card is clickable.  There for a prop` to ` is Router-link object that enables user navigation to the intended route. Think of the current channel cards for example,  We are trying to lean on this [guidelines ](https://inclusive-components.design/cards/)as much as we can

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        Horizontal Layout
+      </h3>
+      <p>
+        It's important to note that besides the mandatory   <DocsInternalLink href="/KCard#prop:title">
```
I think this paragraph could go in a different section before this section, as this doesnt feel part of the "Horizontal Layout" section.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        and
+        <DocsInternalLink href="/KCard#prop:thumbnailDisplay">
+          <code> thumbnailDisplay</code>
+        </DocsInternalLink>  properties. Additionally, the order of slots is not strictly enforced. For instance,
```
I didn't understand this explanation of the order of the slots and the `title`, `aboveTitle`. I think in a previous section we could indicate what our definition of Title is and the difference between `aboveTitle`, `title` prop, `title` slot, `belowTitle`.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        </DocsIternalLink> slot if the latter is not used.
+      </p>
+      <p>
+        The thumbnail's appearance depends on two factors: the size of the original image and the chosen
```
I think we can introduce thumbnails and its appearance in a different section and introduce the thubmnail-related props "thumbnailDisplay", "thumbnailSrc", "thumbnailScaleType". But I dont think we need to introduce this inside the `Horizontal` section, because this is not specific to the horizontal layout and can also be used with the vertical layout

AllanOXDi replied:
That's a good point

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+      </p>
+      Here are a few examples...
+
+      <p>
```
I think this should be a heading to keep consitency with the vertical layout examples

AllanOXDi replied:
Good catch!

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        </div>
+      </DocsShow>
+
+      <h4>
```
Note that in the vertical layout subheadings we are using h5, and here h4

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+          </code>
+        </DocsInternalLink> is not provided we will end something like this
+      </p>
+      <DocsShowCode language="html">
```
Here we are not showing any visual example:
![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/bacfb705-454b-4e6a-af24-29faecc08879)

AllanOXDi replied:
I am thinking that's is the visual example below it- No?  I want to show few visual example.  Otherwise, the  documentation page will get cultured and I have not figured out a way to handle it.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+        </div>
+      </DocsShowCode>
+      <p>
+        And if the layout is set  Horizontal
```
I see diferent layuots between this example and [this one](https://github.com/learningequality/kolibri-design-system/pull/625/files#diff-afca5902f76934ce03daa69d66c65def4f84818a26273d3dff10bd122efa56a5R301) but the description of "horizontal layout" and "small thumbnailDisplay" are the same.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+            :headingLevel="2"
+            title="How to get started with Kolibri"
+            layout="horizontal"
+            thumbnailDisplay="none"
```
What is the difference between an horizontal and vertical layout if we dont have thumbnails in neither?

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,359 @@
+<style lang="scss" scoped>
+
+  em {
+    font-style: normal;
```
With this rule we would be using the `em` tag as a `strong` tag isnt it?

AllanOXDi replied:
No, rule does not actually make the `em `tag function exactly like the `strong` tag.
I think` em ` indicate emphasis or stress within a sentence. Can be rendered as italic, bold, or underline depending on browser and user styles.

`strong ` indicate strong importance of a word or phrase, typically rendered as bold.

AlexVelezLl commented on lib/KCard/BaseCard.vue:
```diff
@@ -0,0 +1,127 @@
+        :to="to"
+      >
+        <KTextTruncator
+          v-if="title !== null"
```
I think this title !== null check should be at the level of the header
```vue
    <component
      v-if="title !== null"
      :is="headerLevel"
    >
```

Because if we dont pass a Title we would be still rendering the header tag and the `router-link` even thought its invisible because we dont have text inside it.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+       * @param {String} value - The layout value.
+       * @returns {Boolean} - True if the value is not empty, false otherwise.
+       */
+      layout: {
```
I think we sould also validate the layout values, and include them in the jsdocs for implementers to know which value this prop can have.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+      >
+        <KImg
+          :src="thumbnailSrc"
+          :isDecorative="true"
```
Is this allways true? Are all the images on a card decorative? Or maybe in the future we could support a `thumbnailAlt`?I'm just asking because I find it curious.

AllanOXDi replied:
I think not all images on a card are always decorative.  From my understanding the `:isDecorative `property used to indicate to assistive technologies (screen readers) whether the image provides essential information or not. Screen readers might skip decorative images to improve reading efficiency for visually impaired users.

Your intuition is right. It's very common to have a descriptive text alternative (alt text) for informative thumbnails. Including a `thumbnailAlt` property would allow us to provide this description, which would be beneficial for visually impaired users as well. https://www.w3.org/WAI/tutorials/images/decorative/

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+  >
+    <template #default>
+      <aside
+        v-if="layout === 'vertical' && thumbnailDisplay !== 'none'"
```
We should avoid burned strings in the code and instead use constants to avoid any typos and make it easier to maintain in the future if we want to change the string.

So we can maybe declare constants:

```javascript
const CardLayout = {
  VERTICAL: "vertical",
  HORIZONTAL: "horizontal",
};

const CardThumbnailDisplay = {
  NONE: "none",
  SMALL: "small",
  LARGE: "large",
};
```

AllanOXDi replied:
I am afraid to take this approach  because  a couple of things could go wrong  here with the` CardLayout` and `CardThumbnailDisplay` typos in the enum values (e.g., "vertival" instead of "vertical") could lead to unexpected behavior or errors.
If we accidentally use a misspelled value (e.g., CardLayout.VERTICLE), it wouldn't match any defined value in the enum

Another thing I am thinking of is that that  it only cover basic layout and thumbnail display options, it only define string values.  And If we need to associate additional data with each layout or display option this  might not be sufficient.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+            <!-- @slot Places content to be placed above the title.. -->
+            <slot name="aboveTitle"></slot>
+            <!-- @slot for the title content -->
+            <slot name="title"></slot>
```
I think this title slot is an alternative to the title prop. So I think we will probably want to make this part of the BaseCard component perhaps?

AllanOXDi replied:
That would not be a good idea. I think `BasCard` is to enforce consistent structure and accessibility for cards across  our product and Isolates basic accessibility requirements from potential changes in the `KCard` component,

MisRob replied:
`BaseCard` is meant to have the validation _"Contains validation to check that title prop or slot has been used, throwing an error if neither is provided"_ (see acceptance criteria here https://github.com/learningequality/kolibri-design-system/issues/529) so I think the slot will need to be present in the `BaseCard`. It will add some clarity as well. This is part of the enforcement that `BaseCard` does - it needs to ensure that title is provided, and that can happen either via the prop or the slot.

@AllanOXDi you too have made a good point and regarding that, we need to ensure that `BaseCard` shouldn't have any styles or layout logic related to how the title appears.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+    :titleLines="titleLines"
+    :to="to"
+  >
+    <template #default>
```
The logic in general is a bit confusing, there are quite a few v-ifs nested in various places and there is repeated code and there is no clear flow of what happens in each case.

I think we could use `display: flex`, and `flex-direction` to play with the layout a bit. Something to keep in mind is that the content will always be the same, what changes is the layout, so we could have something like:

```Vue
<BaseCard>
  <div class="display-flex" :style="wrapperStyle">
    <aside v-if="thumbnailDisplay!='none'">
      <KImg ... :style="imageStyle">
    </aside>
    <div>
      <slot #default>
    </div>
  </div>
  <div class="footer" v-if="$slots.footer">
    <slot #footer>
  </div>
</BaseCard>

...

<script>
...
wrapperSyle() {
  if (layout === "vertical") {
    return { flex-direction: "column" }
  }
  if (layout === "horizontal" && thumbnailDisplay === "small") {
    return { flex-direction: "row-reverse" } // To kep the image on the right
  }
  if (layout === "horizontal" && thumbnailDisplay === "large") {
    return { flex-direction: "row" } // To keep the image on the left
  }
}
</script>
```

In this way we do not repeat so many KImg, nor the content and we only play with the styles of the image, of the content depending on each case.

MisRob replied:
[required] I second that this would be a valuable improvement and suggest we do it right in the first iteration as it will make any further work easier

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+      },
+      /**
+       * Controls the display of the thumbnail image.
+       * Options: 'none' (default), 'block', or 'inline'.
```
The options we say here are 'none', 'block', 'inline'. But the ones we use in the code are 'none', 'small' and 'large'.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,226 @@
+      isVerticalLayout() {
+        return (
+          this.layout === 'vertical' ||
+          (this.layout === 'horizontal' && this.thumbnailDisplay === 'large')
```
Does this means that an horizontal layout with large thumbnail behaves like a vertical layout?

AllanOXDi replied:
Yes, my reasoning behind this  is that a large thumbnail in a horizontal layout might visually occupy a significant portion of the space, creating a layout that resembles a vertical one where elements are stacked.

AlexVelezLl commented on lib/KCard/BaseCard.vue:
```diff
@@ -0,0 +1,127 @@
+
+  .spacing {
+    padding: 1em;
+    color: black;
```
Here we are setting a burned color, instead of the $themeTokens ones.

AlexVelezLl reviewed (changes requested):
Hi @AllanOXDi! I am glad that our KCard is already taking shape :tada:. I noticed a few things, especially in the documentation, and in the KCard code structure. Let me know any thoughts of any of the comments :open_hands:.

MisRob commented on docs/assets/vertical-layout.png:
[required] I believe this is unused file that can be removed

MisRob commented on lib/styles/trackInputModality.js:
```diff
@@ -25,6 +25,7 @@ function setUpEventHandlers(disableFocusRingByDefault) {
     '[role=textbox]',
     'a',
     'button',
+    '[data-focus=true]',
```
[required] Could you add a brief comment on the motivation behind this and mention its link to `KCard`'s focus?

AlexVelezLl replied:
Seems like this is still missing

MisRob reviewed (commented):
Wonderful work overall @AllanOXDi, and it's great that a11y QA passes, that was the most tricky thing!

I am posting some first feedback. I haven’t reviewed yet everything - will follow up gradually.

MisRob commented:
[required] `title` slot is to be an alternative to the `title` prop, however now the slot now renders in addition to the `title` slot and is not showing in a correct place. For example, for

```
<KCard
    :to="{ name: '/' }"
    :headingLevel="2"
    layout="vertical"
  >
    <template #title>
      Card title
    </template>
</KCard>
```

I see the title text not being rendered as heading:

![Screenshot from 2024-05-18 13-22-17](https://github.com/learningequality/kolibri-design-system/assets/13509191/501b3b6b-09cf-430e-bf1a-0e5690360f73)

Could you adjust the behavior to follow the expected behavior in all of the following cases? Note that in one or two cases it works well already, but not in all of them - including all for clarity.

| | Expected |
| --- | --- |
| only `title` prop ```<KCard title="Card title">``` | ![Screenshot from 2024-05-18 13-12-14](https://github.com/learningequality/kolibri-design-system/assets/13509191/7ea61e1c-34f2-4a54-9d06-5d08ae1cc12c) |
| only `title` slot ```<KCard><template #title>Card title</template></KCard>``` |  ![Screenshot from 2024-05-18 13-12-14](https://github.com/learningequality/kolibri-design-system/assets/13509191/7ea61e1c-34f2-4a54-9d06-5d08ae1cc12c) |
| both `title` prop and `title` slot ```<KCard title="Card title from the prop"><template #title>Card title from the slot</template></KCard>``` - give preference to the `title` prop and ignore the slot | ![Screenshot from 2024-05-18 13-20-30](https://github.com/learningequality/kolibri-design-system/assets/13509191/9c39c8f4-a523-49d2-bb62-16e66106d519) |

MisRob commented:
[required] `title` slot is to be an alternative to the `title` prop, however now the slot renders in addition to the `title` slot and is not showing in a correct place. For example, for

```
<KCard
    :to="{ name: '/' }"
    :headingLevel="2"
    layout="vertical"
  >
    <template #title>
      Card title
    </template>
</KCard>
```

I see the title text not being rendered as heading:

![Screenshot from 2024-05-18 13-22-17](https://github.com/learningequality/kolibri-design-system/assets/13509191/501b3b6b-09cf-430e-bf1a-0e5690360f73)

Could you adjust the behavior to follow the expected behavior in all of the following cases? Note that in one of them it works well already, but not for all - including all for clarity. There's also an additional issue in the screenshots below regarding the placement of the `aboveTitle` but I believe it is already being addressed.

| | Expected |
| --- | --- |
| only `title` prop ```<KCard title="Card title">``` | ![Screenshot from 2024-05-18 13-12-14](https://github.com/learningequality/kolibri-design-system/assets/13509191/7ea61e1c-34f2-4a54-9d06-5d08ae1cc12c) |
| only `title` slot ```<KCard><template #title>Card title</template></KCard>``` |  ![Screenshot from 2024-05-18 13-12-14](https://github.com/learningequality/kolibri-design-system/assets/13509191/7ea61e1c-34f2-4a54-9d06-5d08ae1cc12c) |
| both `title` prop and `title` slot ```<KCard title="Card title from the prop"><template #title>Card title from the slot</template></KCard>``` - give preference to the `title` prop and ignore the slot | ![Screenshot from 2024-05-18 13-20-30](https://github.com/learningequality/kolibri-design-system/assets/13509191/9c39c8f4-a523-49d2-bb62-16e66106d519) |

MisRob commented:
[required] One important area to address here is to check the resulting UI against [the "Updated Cards Specs Metadata" designs](https://www.figma.com/design/1tQQ0Sw1yDjzO66hyjLVPM/Kolibri-UX-Fixes?node-id=1868-55352) and make sure that all the logic and styles that `KCard` is responsible for aligns to them (while leaving out areas that `KCard` is not responsible for, such as content of slots like icons, buttons, etc.) Few examples:

|Design | Implementation | 
| -------- | -------------------- |
| ![Screenshot from 2024-05-18 14-23-11](https://github.com/learningequality/kolibri-design-system/assets/13509191/e9113bf4-a92a-44c3-81fb-dfa9d6bee3bf) | ![expected-1](https://github.com/learningequality/kolibri-design-system/assets/13509191/9ad81f6e-bb63-4e2a-a40e-b439b6e9e8d0) |
| ![Screenshot from 2024-05-18 14-35-32](https://github.com/learningequality/kolibri-design-system/assets/13509191/a5d389da-52ce-41ee-9919-fefdfd31e0a9) | ![expected-2](https://github.com/learningequality/kolibri-design-system/assets/13509191/2bc77534-718d-4981-867d-7d395a0634ec) |
| ![Screenshot from 2024-05-18 14-42-42](https://github.com/learningequality/kolibri-design-system/assets/13509191/a3cd0e41-6d1b-4cf0-852d-bfd6d66e5c13) | ![expected-3](https://github.com/learningequality/kolibri-design-system/assets/13509191/b0feb186-8c01-4763-9c6e-1f591366d66e) |

These are just a few places I noticed, and is not meant to be a full list. Could you please take some time to test all available layouts and compare with designs? Placement, font sizes, padding/margins, and pretty much anything we can do to follow designs closely. Figma will provide you with all information and useful measurement tools, such as

<img width="300" src="https://github.com/learningequality/kolibri-design-system/assets/13509191/491f84fd-b7ee-4ecc-b99a-8e17f48916fc">

Note that letterboxing should be preserved though even though it's not showed on the designs. The image on the designs corresponds to the whole thumbnail area (possibly letterboxed) in the final implementation.

MisRob commented:
@AllanOXDi There are some acceptance criteria from the issues this PR is meant to close that appear not yet be implemented or have unclear status. Could you please read through all related issues, particularly "Markup/style requirements" and the "Acceptance criteria" sections and resolve them or mention where things are if there are any blockers? Some examples: "_Supports RTL_", "_Even though the whole area is clickable, text content is still selectable_", "[documentation] _contains guidance on the headingLevel to notify developers that they need to choose a correct level based on context_", and perhaps few more.

I know it's lots of work, but there's time to finish it and would suggest we do so in this PR since it's not draft like the first Studio PR. I think it will be easier than opening many follow ups since there's no pressure around merging. This will prepare robust ground for further work and help with implementing card grids as well. If you needed to chat around how to organize work or encountered some blocking issues, we'll have time to chat :)

MisRob commented on lib/KCard/__test__/KCard.spec.js:
```diff
@@ -0,0 +1,76 @@
+
+  //testing slots
+  it('renders slotted content', () => {
+    expect(wrapper.find('[slot="aboveTitle"]').exists()).toBe(false);
```
Did you mean to test for the presence of the slot? If so, you'd need to say `toBe(true)`. The test, as implemented now, confirms that the slot _doesn't exist_ rather than that it does.

I think that lines 65-68 are sufficient to confirm that slot content gets displayed.

AllanOXDi replied:
I was trying to make sure that the card does not emit data by default and only emits data when it receives focus.

MisRob commented on lib/KCard/__test__/KCard.spec.js:
```diff
@@ -0,0 +1,76 @@
+  });
+
+  //testing focus and hover events
+  it('renders data on focus', () => {
```
Could you please elaborate a bit on what this scenario is supposed to test?

AllanOXDi replied:
I was trying to make sure that the card does not emit data by default and only emits data when it receives focus.

MisRob commented on lib/KCard/__test__/KCard.spec.js:
```diff
@@ -0,0 +1,76 @@
+    expect(emittedData).toBeUndefined();
+  });
+
+  it('emits data on hover', () => {
```
Could you please elaborate a bit on what this scenario is supposed to test?

AllanOXDi replied:
I was trying to make sure that the card does not emit data by default and only emits data when it receives focus.

MisRob replied:
What kind of data is it expected to emit on receiving the focus?

MisRob commented:
@AllanOXDi I think I've finished the first round. At this point, I focused rather on higher-level issues. I will review one more time after this feedback as well as feedback from @radinamatic and @AlexVelezLl is addressed since there will be some important changes. Thanks a lot for this work, it's shaping up really great!

AllanOXDi committed:
change the path to lowercase

AllanOXDi committed:
change the thumnail  logic

AllanOXDi committed:
revert semantic elements

AllanOXDi committed:
used css to display control card content display

AllanOXDi committed:
used themetoken color

AllanOXDi committed:
title displays in the correct section'

AllanOXDi committed:
added a comment to the data-focus=true

AllanOXDi committed:
removed unused image file

AllanOXDi committed:
added conditionalized image radius

AllanOXDi committed:
cleaned overriding title text

AllanOXDi committed:
changes test suit case

MisRob commented:
For the reviewers, just a note that resolving the first PR feedback is still in progress, so the time for re-review will come later.

AllanOXDi committed:
adds comments

AllanOXDi committed:
returned the header code to the basecard component

AllanOXDi committed:
checking content display

AllanOXDi committed:
implement mousedown and mouseup for selecting text

AllanOXDi committed:
Error thrown when no title prop and title slot provided

AllanOXDi committed:
prevents routing when copying text

AllanOXDi committed:
headerlevel value check

AllanOXDi committed:
added checks on props as per the specifications

AllanOXDi committed:
thumbnail scale validation

AllanOXDi committed:
fixes console errors on the kcard documentation page

AllanOXDi committed:
cleanedup the slot and prop description text on the kcard documentation

AllanOXDi committed:
adds more description on the kcard documentation

AllanOXDi committed:
fixes having wrong layouts passed as examples on the documentation page

AllanOXDi committed:
grammar clean up

AllanOXDi committed:
test <li> element

AllanOXDi committed:
test correct accessibility structure

AllanOXDi committed:
adds the footer to the extreme bottom in the horizontal layout and small thumbnail

AllanOXDi committed:
adds a fontsize to the above title

AllanOXDi committed:
update the card

AllanOXDi committed:
work in progress

AllanOXDi committed:
rabese to develop

AllanOXDi committed:
removes the border radius from vertical layout kimg

AllanOXDi committed:
redesigned card content

AllanOXDi committed:
adds spacing

AllanOXDi committed:
adds a border radius

AllanOXDi committed:
adds playground sample UI for all layouts

AllanOXDi committed:
lint

AllanOXDi committed:
lint

AllanOXDi committed:
fixes js failure

AllanOXDi committed:
fixes doc page

AllanOXDi committed:
rebase

AllanOXDi committed:
lint

AllanOXDi committed:
lint

MisRob commented:
Note that RTL will be done as part of https://github.com/learningequality/kolibri-design-system/issues/665 rather than in this PR

MisRob commented on lib/KCard/BaseCard.vue:
```diff
@@ -0,0 +1,144 @@
+    cursor: pointer;
+    border-radius: 0.5em;
+    outline-offset: -1px;
+    transition: border-color $core-time ease;
```
I was just doing KDS shadows motion audit for designers when I noticed the transition doesn't seem to work in the card. I think we'd need to use `box-shadow` instead of `border-color`? Since that's the style that dropshadows as defined here modify https://github.com/learningequality/kolibri-design-system/blob/d784b830967d810ae88a42e207b755bbe5bc2318/lib/styles/definitions.scss#L59

AllanOXDi committed:
renders the corect the markup

AllanOXDi committed:
fixes verticle layouts

AllanOXDi committed:
uses flex to display kcard content

AllanOXDi committed:
fixes indentation errors

AllanOXDi committed:
conditionalize image slot appearance

AllanOXDi committed:
adds background color to the card

AllanOXDi committed:
fixes kcard responsiveness

AllanOXDi committed:
rebasing

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,416 @@
+    </DocsPageSection> <DocsPageSection title="Usage" anchor="#usage">
+      <p>
+        <code>KCard</code> is a versatile base component for creating various card types.
+        It can be configured by adding product-specific content and logic to generate distinct card components.
```
This seems very similar to the text in the introduction section. I would suggest to keep this text as introduction instead of the current paragraph, as it seems more general.

AlexVelezLl commented on docs/pages/kcard.vue:
```diff
@@ -0,0 +1,416 @@
+      <DocsShow block>
+        <div>
+          <KCard
+            :to="{ name: '/' }"
```
I think we could add the redirect to the same kcard page here: `:to="{ name: 'kcard' }"` to avoid the redirection error in the documentation when we click on the card:
![Screenshot from 2024-07-09 13-24-16](https://github.com/learningequality/kolibri-design-system/assets/51239030/6970d591-4ecb-425b-a648-4d1f78217995)

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+          (this.thumbnailDisplay === 'small' || this.thumbnailDisplay === 'large')
+        ) {
+          return {
+            width: '50%',
```
This value could be calculated dinamically, since because of this restriction, we can see too much padding in the card:
![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/3d35ca3a-be62-4dc0-b40f-d8c1f1d6f3bc)

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+  >
+    <template #title>
+      <!-- @slot Optional slot section containing the title contents, should not contain a heading element. -->
+      <div class="title-slot">
```
It seems like we are not using the title slot of the basecard, since we dont have the heading level set on the title text: 
![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/4a86e647-0544-4463-8a4a-4ddf74631765)

We would need to add the <slot name="title"/> here to pass the title slot of the KCard to the title slot of the Base Card

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+  }
+
+  /deep/ .base-card-heading {
+    order: 2;
```
It seems we have a little error here. Because of this order: 2 style, we are having the title slot at the end of the card, and since we are not using the title slot of the base card, it appears as an empty space at the bottom of a vertical card:

![image](https://github.com/learningequality/kolibri-design-system/assets/51239030/ece413d9-1681-4851-b14d-bd9bbbbd683b)

AlexVelezLl commented on lib/KCard/BaseCard.vue:
```diff
@@ -0,0 +1,154 @@
+      },
+      onMouseUp() {
+        const mouseUpTime = new Date().getTime();
+        if (mouseUpTime - this.mouseDownTime < 200) {
```
I think it would be nice to have a comment here explaining that the reason for this behaviour is to allow selecting text inside the card.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+          class="above-title-style"
+          :class="{ 'above-title-horizontal': layout === 'horizontal' }"
+        >
+          <slot name="aboveTitle"></slot>
```
@MisRob I wonder if these "aboveTitle", "belowTitle" slots should also be present on the  BaseCard component? I dont see these slots in the BaseCard specs. But we dont have a way of control these slots here in the KCard to make them appear above or below the title, since it is the BaseCard who has the responsibility of rendering the title, so probably its also the BaseCard the responsilble for rendering the belowTitle and aboveTitle.

MisRob replied:
@AlexVelezLl There's a way to control them via styling with flex order which is needed for more reasons anyways - primarily because the placement of slots in the code really shouldn't change the resulting markup structure. However, even though it's technically valid and possible, it creates some complexity when `KCard` needs to do some styling in the `BaseCard`, so I hear you. If the separation of `KCard` and `BaseCard` brings more cons than pros, an alternative that would resolve all this would be to merge these two components into a single `KCard` component. I'm open to this and leave to @AllanOXDi's judgement.

MisRob replied:
Note that even if we merged those two components though, we'd still need to show the content of `aboveTitle` below the title in the markup and achieve placing it above the title by means of CSS rather than HTML. This is a standard a11y friendly coding technique and a core feature of the card ensuring that screen readers announce it as expected.

So the question of merging vs not would really be about cross-component vs one-component styling. In both cases we'll still be using `flex-order`.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+        <div
+          data-testid="aboveTitle"
+          class="above-title-style"
+          :class="{ 'above-title-horizontal': layout === 'horizontal' }"
```
I dont see any CSS rules defining none of the classes: above-title-horizontal, title-horizontal, below-title-horizontal, so we will need to remove them from the nodes also.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+          <slot name="belowTitle"></slot>
+        </div>
+        <div
+          class="footer"
```
I think we could add a check here, and render this node just if there is a `$slots.footer`

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,314 @@
+    FIT_XY: 'fitXY',
+  };
+
+  function isValidScaleType(value) {
```
I dont think we should do this validation logic here, since we already do it inside the KImg component, and this create duplicated code, and we have to keep in mind update both places if we need to add a new ScaleType for example

MisRob replied:
Agreed

MisRob replied:
Agreed

AlexVelezLl commented on lib/KCard/__test__/KCard.spec.js:
```diff
@@ -0,0 +1,110 @@
+  return mount(KCard, { router, propsData, slots, localVue });
+}
+
+describe('index.vue', () => {
```
The name of this Test Suite should be the same as the component we are testing, to more easily identify which component each test belongs to.

AllanOXDi replied:
I did not get you here

AlexVelezLl replied:
Oh I was talking about renaming the describe block to 
```js
describe('KCard', () => {
```

To see in the  tests output that it is testing the KCard component. This is the output right now, and as you can see index.vue is not very descriptive (especially because this is logged among a lot of other tests).

```
  index.vue
    ✓ renders passed header level (20 ms)
    ✓ renders the correct accessibility  structure when title slot is passed (14 ms)
    ✓ renders the correct accessibility structure when title pass as props (6 ms)
    ✓ should not navigate on long click (508 ms)
    ✓ should navigate on quick click (109 ms)
    it renders slotted content
      ✓ renders slotted content with belowTitle slot (6 ms)
      ✓ renders slotted content with footer slot (6 ms)
```

AlexVelezLl commented on lib/KCard/__test__/KCard.spec.js:
```diff
@@ -0,0 +1,110 @@
+    it('renders slotted content with belowTitle slot', () => {
+      const wrapper = makeWrapper({
+        propsData: { to: { path: '/some-link' }, title: 'sample title ' },
+        slots: {},
```
I think we missed passing slots here. Here we have the intention of test slotted content, but we are not passing anything to the slots prop.

AlexVelezLl reviewed (commented):
Looks amazing!! This is already super polished and looks super good. I just left some minor things I noticed that could add value to the implementation <3

MisRob reviewed (dismissed):
Thank you for updates, @AllanOXDi, and @AlexVelezLl for re-review. I won't re-review everything right now since I can see there's lots of feedback from @AlexVelezLl. However, I will request one change - there are important regressions in regard to card title, when passed both via slot and via prop. This is important to have in place since it's why we're building the card.

This is the expected output structure markup of a card, under any circumstances:

![Screenshot from 2024-07-10 17-44-07](https://github.com/learningequality/kolibri-design-system/assets/13509191/0db2ac7a-d686-4609-b92d-e54519033d5b)

At this point, when the title is passed via slot, the markup structure is

![Screenshot from 2024-07-10 17-37-50](https://github.com/learningequality/kolibri-design-system/assets/13509191/3c8998a6-e1a1-4ada-bbe0-23be7cfbb9a6)

and when passed via prop, it is

![Screenshot from 2024-07-10 17-39-01](https://github.com/learningequality/kolibri-design-system/assets/13509191/43dad979-3cd6-42b8-9766-2352fd478f59)

Even though the `title-slot` div is there, it's always empty and the title is rendered somewhere else than here in `<a>` which is the main problem here. 

Likely related, the title doesn't show at all when passed as prop:

![Screenshot from 2024-07-10 17-46-13](https://github.com/learningequality/kolibri-design-system/assets/13509191/eff99e10-db1a-4540-8575-4dbfa8a10184)

It may help you to return more examples to the playground page, for more layout modes of the card, and also for when title is passed to the card in both these ways. Please make sure that in these cases, the card satisfies markup and visual requirements.

In regard to minor feedback, I think it'd be fine to open follow-ups it if helps. Please let me know if you needed anything.

AllanOXDi committed:
removes unused test, functions  and css classes

AllanOXDi committed:
passes slot at testing

AllanOXDi committed:
fixes comments

AllanOXDi committed:
card content placed at their position

AllanOXDi committed:
spacing

AllanOXDi committed:
stablizes vertical layout display levels

AllanOXDi committed:
vertical small image spacing

AllanOXDi committed:
validation of the tile slot/prop

MisRob committed:
Fix and cleanup title content and styles

- Remove duplicate 'KTextTruncator's and make
  sure that the title gets rendered when passed
  via prop
- Make title styles consistent no matter of whether
  the title is provided via prop or slot
- Cleanup duplicate title styles
- Several minor style organization improvements

MisRob committed:
Remove .pytest_cache files

MisRob committed:
Add more documentation to BaseCard

MisRob committed:
Finish text selection feature

- cleans up unnecessary handlers
- adds new logic that is needed
  for the feature to work
- contains minor renaming of other handlers
  just for consistency.

MisRob committed:
Move a11y related validators to BaseCard

It's BaseCard that is utilizing heading level
and title, and therefore that's the place to ensure
that these data are provided in the expected format.

This also removes pieces of validation there weren't
needed. For example, Vue takes care of checking
non-empty value for required props, so we don't
need to do it.

MisRob commented:
Hi @AllanOXDi, I started the final round of review. Overall it feels much more robust now, layouts look great, and I really appreciate you took time to prepare cards preview similar to the cards we will implement in Kolibri. I am sure this will make all further work in products smooth. Thank you for all hard work :)

As we chatted earlier, it'd be great to have this merged before you're back online and open follow-ups if there's some more work. I will open a few, but for some changes, it's a bit easier for me to push them right away rather than explaining in new issues, and I also hope it will help with other work we have in progress in Kolibri and Studio that you can return to when you are back. I hope you don't mind - it's done with the intention of support. I'm trying my best to describe reasons for changes in the commit messages, and will be available to talk about each of them. The main areas updated I considered important

- Fixes for some last issues with the title slot
  - Motivation: User-facing 
- Consistency of the title styles no matter if title provided via prop or slot
  - Motivation: User-facing  
- Text selection feature completion
 - Motivation: User-facing  
- Some last styling touches
  - Motivation: User-facing  
- Re-organization of styles in `KCard` regarding various layouts
  - Motivation: This is really not so important for users, however I believe it will help a lot with future development, and it also helped me with some styling updates. I think we had this done already as response to @AlexVelezLl's first review, but seems to get lost with big styles overhaul after fixes to markup structure.

(You will also see some documentation and renaming mixed in my commits and that's just me nitpicking and doing little improvements when I see an opportunity as I'm in the code already, nothing really important, and in majority of cases, I wouldn't be posting to review :)

After I am done, I will ask @AlexVelezLl to review my updates, and then we will merge. Thanks both and talk to you soon.

MisRob committed:
Fix some tests, follow naming conventions

Also includes minor improvements of docstrings
and tests structure.

MisRob committed:
Several styling fixes, robust styling approach, re-organization of styles

Relatedly
- Removes 'aside' element (in some layouts it doesn't correspond to its meaning)
- Removes 'titleLines' prop since this is (and needs to be) controlled by layout
- Removes /deep/ selectors in favor of using 'headingStyles' prop (the
implementation dependet on nested selectors with /deep/ in the
middle, however it only seemed to work by chance when missing scss language
config. For <style scoped lang="scss"> it caused the page compilation
break.

MisRob committed:
Make placeholders work

MisRob committed:
Resolve changelog conflict

MisRob committed:
Fix linter errors

MisRob commented:
Hi @AllanOXDi, I finished the final review. Overall it feels much more robust now, layouts look great, and I really appreciate you took time to prepare cards preview similar to the cards we will implement in Kolibri. I am sure this will make all further work in products smooth. Thank you for all your hard work :)

As we chatted earlier, it'd be great to have this merged before you're back online and open follow-ups if there's some more work. I will open a few, but for some changes, it's a bit easier for me to push them right away rather than explaining in new issues, and I also hope it will help with other work we have in progress in Kolibri and Studio that you can return to when you are back. I hope you don't mind - it's done with the intention of support, and hope it aligns well with all we talked about together. I'm trying my best to describe reasons for changes in the commit messages, and will be available to talk about each of them. The main areas updated I considered important (cc @AlexVelezLl for review):

**User-facing:**
 - Fixes for some last issues with the title slot
 - Consistency of the title styles no matter if title provided via prop or slot
 - Improvements to styles robustness and content tolerance,  and some last styling touches
- Ensuring that thumbnails can display placeholders via "thumbnailPlaceholder" slot when thumbnail is not available

**Other:**
- Re-organization of styles in `KCard` regarding various layouts so that styles belonging to particular layouts are grouped together
  - This is really not so important for users, however I believe it will help with future development, and it also helped me a lot with some styling updates I am mentioning above. I think we had this done already to some extent as response to @AlexVelezLl's first review, but seems to have gotten lost with big styles overhaul after fixes to markup structure. It's lots of changes but rather organizational with a few fixes applied, the foundational techniques you put in place stay the same.
 
(You will also see some documentation and renaming mixed in my commits and that's just me nitpicking and doing little improvements when I see an opportunity as I'm in the code already, nothing really important, and in majority of cases, I wouldn't be posting to review :)

After I am done, I will ask @AlexVelezLl to review my updates, and then we will merge. Thanks both and talk to you soon.

MisRob commented:
Would you please give the components one last look, @AlexVelezLl? Please also see the playground page that simulates more use-cases now. I will delete it before merging, but keep a copy somewhere for testing (this will be perfect first task for automated visual tests :)

Also please note that the following will be resolved in follow-ups (some already open, and I will open later remaining ones). If there's any feedback from you that's not resolved and not blocking, please mention.

- Clarify how cards will behave in a grid regarding their height with designers, e.g. should the cards height be always the same or should the row adjust to the highest card (current implementation is my best guess)? Based on that, determine if we need to be able to override cards heights/widths, remove spaces occupied by empty slots, etc. (I will do this as part of grid work)
- Some more guidelines in documentation for frequent tasks (probably me)
- Vertical/horizontal layout with no thumbnail (me or Allan)
- Adding constants for `layout` and `thumbnaiDisplay` and related prop validators (Allan)
- Implementing RTL support (Allan)

AlexVelezLl commented on lib/KCard/BaseCard.vue:
```diff
@@ -0,0 +1,204 @@
+        // If the time difference is greater than or equal to 200 milliseconds,
+        // it means that the mouse button was pressed and held for a longer time,
+        // which is not typically interpreted as a click event. Do not navigate
+        // in such case.
```
I see value in adding that this allows us select text for example.

MisRob replied:
Ah yes, it was couple of lines above - will make sense to have it here together. Updated, thank you.

AlexVelezLl commented on lib/KCard/index.vue:
```diff
@@ -0,0 +1,393 @@
+        data-test="aboveTitle"
+        class="above-title"
+      >
+        <slot name="aboveTitle"></slot>
```
I think we should document these slots as well.

MisRob replied:
Ah yes, thank you. Done

radinamatic commented:
I took a look at the latest version of the playground and tested with NVDA in Windows. 

One thing popped up as a concern, but again I'm not sure if it's determined by this base card implementation, or will it be further _injected_ by other products. Namely, the color contrast of the channel name (is that in the ``aboveTitle`` slot?) is insufficient:


<img width="841" alt="2024-07-19_16-57-29" src="https://github.com/user-attachments/assets/e0bffb74-6dd6-4bc9-b7cc-8ffcf90c7c72">



Similarly, even though I understand what was mentioned earlier, that:

> products that will be responsible for adding buttons, context menus, etc. and ensuring it doesn't break a11y.

I just wanted to reiterate 🙂 , that we need to pay attention to offer full and comprehensive information about those element in the footer. For example, this is the full readout of one of the cards by NVDA:

```
clickable  
<Title>  
link    
heading    level 2  
<Below title>
PracticeShort Activity  20
```

All above looks Ok, just 2 comments regarding the last line:
* a separation is missing between the type of resource and the duration 
* [``progress``](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress#accessibility) element needs a role and a label, otherwise the screen reader outputs just the number (20) without any context

I know this might not be in the scope of the present PR, but it cannot hurt to repeat it 😉

AlexVelezLl reviewed (approved):
This LGTM! Code looks good and well documented, documentation looks very informative and easy to follow, and visual examples in the playground looks great! I think is ready to go! Just left a super minor comments about docs. 

Thank you @AllanOXDi for all your consistent hard and great work here!! and @MisRob for your great guidance! Will leave the final merge to @MisRob once playground is cleaned and QA passes.

MisRob committed:
Move comment

MisRob committed:
Document slots

MisRob committed:
Fix value type

MisRob committed:
Remove obsolete styles

!important was needed in times when 'thumbnail'
was applied directly to KImg but this was later
changed and is now applied to its wrapping div
so it's not needed anymore.

MisRob committed:
Cleanup playground

MisRob commented:
Thank you @radinamatic, I appreciate it. Yes, these are all the things that consuming apps need to ensure as they will inject it. I'd like us to have some guidance into the documentation to clarify for developers what we need to take care of when implementing cards on top of this component vs what is it already taking care of. Also, as soon as we have first few Kolibri card components implemented and a11y tested, we could use the way they look for the documentation examples to reflect this all a bit better.

MisRob commented:
^ We will keep working on this all gradually as part of the project, this component is a core step, and also a beginning of larger body of work in a sense.

MisRob commented:
Thanks for feedback @AlexVelezLl! Merging now.
